### ISSUE TITLE
Integrate  text-to-speech and speech-to-text functionalities

### ISSUE THREAD

**Issue by amakropoulos** [undreamai/LLMUnity#44]

**Comment by ArEnSc** [comment]
please make this an optional package that is separate

**Comment by amakropoulos** [comment]
Yes certainly, it will be possible to attach STT or TTS to the chat functionality but it will not be enabled by default.

**Comment by simoninithomas** [comment]
Hey there 👋 , will you use Sentis for STT and TTS? Or do you have another idea? 

We have some Sentis model on the Hub that are super fast (Tiny Whisper and Jets).

Tiny Whisper: https://huggingface.co/unity/sentis-whisper-tiny
Jets: https://huggingface.co/unity/sentis-jets-text-to-speech

Demo with Whisper: https://singularite.itch.io/jammo-the-robot-with-unity-sentis-whisper-version

**Comment by amakropoulos** [comment]
Hi, thank you for the suggestions!
I need to do a small exploration first, but yes I was thinking to start with your Whisper-Tiny model 🙂.
Ideally I would like to support a range of models e.g. similarly to [whisper.cpp](https://github.com/ggerganov/whisper.cpp) project but need to have it working cross-platform in Unity which is work-in-progress ([link](https://github.com/Mozilla-Ocho/llamafile/issues/17)).

By the way, thanks a lot for your great work on the sharp-transformers ⭐!
I'm using it in the other repo, [RAGSearchUnity](https://github.com/undreamai/RAGSearchUnity), to build a RAG similarity search system!

**Comment by siddhant-bharti** [comment]
Hi @amakropoulos : I want this functionality for a project I am building! Are you planning to add this soon? I can help raise a PR for this functionality too if you are fine with this? Looking forward to hearing from you. Thanks!

**Comment by amakropoulos** [comment]
@siddhant-bharti I'm replying here as well :).
This is the next big feature that I'll work on soon.

@simoninithomas I can't use Jets because it has a cc-by-4.0 license.
The Unity Asset store does not allow packages with licenses that require attribution and I'd like to LLM for Unity to be there as well (p.s. we are live on asset store as of last week :tada: !)

**Comment by amakropoulos** [comment]
This feature is blocked at the moment.
I can't find an open-source library for TTS to integrate that fulfills the following requirements:
- C/C++/C# code without many dependencies
- MIT/Apache 2.0 or any other equivalent license that is open-source and attribution-free
- allow multiple voices

The best solution would be Piper but at the moment has a potential license issue due to to using espeak ([link](https://github.com/rhasspy/piper/issues/455)).

**Comment by Swiftyos** [comment]
Adding TTS and STT functionality would take llamafile to the next level!

**Comment by SubatomicPlanets** [comment]
I also think that piper is the best choice for TTS and whisper for STT. I made a project using the [UnityPiper](https://github.com/mrtrizer/UnityPiper) and [whisper.unity](https://github.com/Macoron/whisper.unity) projects. It works, but it was a bit complicated getting it all to work. I also found [Piper without espeak](https://github.com/gudatr/piper-without-espeak) but I don't know how well it works.

**Comment by 99bits** [comment]
Having this feature is super Amazing and Instead of being blocked by Unity's opensource license requirements.  Could this feature be delivered via Github instead of the unity asset store?

**Comment by amakropoulos** [comment]
I can't spend months to implement it when it can't go to any Unity or commercial product.
If someone wants to use TTS they can have a look at these options instead:
**commercial:** [Overtone](https://assetstore.unity.com/packages/tools/generative-ai/overtone-realistic-ai-offline-text-to-speech-tts-251304), [ElevenLabs](https://elevenlabs.io/)
**open-source (Piper):** [piper.unity](https://github.com/Macoron/piper.unity), [UnityPiper](https://github.com/mrtrizer/UnityPiper)
I haven't personally used any of these, but I know a lot of people that use the commercial ones.

**Comment by amakropoulos** [comment]
interesting, thanks for letting me know.
which models are you using?
